### PR TITLE
refactor: reporting mechanism

### PR DESCRIPTION
**Pull request by leodido**
This PR iterates on the `Reporter` stuff that @fntlnz did.

The goals are:

- give them the awareness of the context in which they can run, or they should error, or they should stop without making the CLI erroring out (eg., when `gh-pull-*` reporters run on push event rather than pull request events)
- encapsulate most of their logic avoiding having to implement it in all our command implementations

This PR preps the ground for #205  (and future others in the same concept space).

---

- [x] I have read the [contributing guidelines](https://github.com/listendev/lstn/blob/main/.github/CONTRIBUTING.md)
- [x] I have followed the [coding guidelines](https://github.com/listendev/lstn/blob/main/docs/coding-guidelines.md)
- [x] I have written unit tests
- [x] I have made sure that the pull request is of reasonable size and can be easily reviewed

### CODE REVIEW

File: cmd/scan/scan.go [hunk 1]
```diff
@@ -151,25 +149,22 @@ The verdicts it returns are listed by the name of each package and its specified
 			}
 
 			for _, r := range scanOpts.Reporter.Types {
+				cs := io.ColorScheme()
```
Comment (fntlnz) [inline cmd/scan/scan.go:152]: this can go outside the for loop

**Review by fntlnz — COMMENTED** [review]
love the refactoring, just a small comment but lgtm for now - waiting for the future changes you will be doing to approve (it's a draft).

🚀

**Comment by leodido** [comment]
image

**Review by fntlnz — APPROVED** [review]

File: pkg/reporter/gh/comment/comment.go [hunk 2]
```diff
@@ -105,3 +114,8 @@ func (r *Reporter) Report(req *request.Report) error {
 
 	return nil
 }
+
+// TODO: do not execute on GitHub push events
```
Comment (leodido) [inline pkg/reporter/gh/comment/comment.go:118]:
```suggestion
// TODO: do not execute on GitHub push events.
```

**Pull request merged**